### ISSUE TITLE
Bug: `no-useless-default-assignment` false positive when `exactOptionalPropertyTypes: true`

### ISSUE THREAD

**Issue by atitcreate** [oxc-project/tsgolint#735]
cd to a temp dir, run:
```bash
npm install --save-dev eslint typescript typescript-eslint oxlint oxlint-tsgolint
npx tsc --init
```
create `eslint.config.mjs`:
```js
import { defineConfig } from "eslint/config"
import tseslint from "typescript-eslint"
export default defineConfig({
  files: ["test.ts"],
  extends: [tseslint.configs.base],
  rules: {
    "@typescript-eslint/no-useless-default-assignment": "error",
  },
  languageOptions: {
    parserOptions: {
      projectService: true,
    },
  },
})
```
create `.oxlintrc.json`:
```jsonc
{
  "categories": {
    "correctness": "off"
  },
  "rules": {
    "typescript/no-useless-default-assignment": "error"
  }
}
```
create `test.ts`:
```ts
function test(options?: { offset?: number }) {
  const { offset = 5 } = { ...options }
  offset.toString()
}
```

### With `exactOptionalPropertyTypes: false`

Modify the generated `tsconfig.json` and set `exactOptionalPropertyTypes: false`
```bash
npx eslint test.ts
# No error 

npx tsc
# No error 

npx oxlint --type-aware test.ts
# No error 
```

### With `exactOptionalPropertyTypes: true`

Modify `tsconfig.json` and set `exactOptionalPropertyTypes: true`
```bash
npx eslint test.ts
# No error 

npx tsc
# No error 

npx oxlint --type-aware test.ts
# False positive:
#   typescript-eslint(no-useless-default-assignment): Default value is useless because the property has type `number` (not nullish). This default assignment will never be used.
```

Then, if applying the fix provided by oxlint, tsc will report error
```bash
npx oxlint --type-aware --fix test.ts
# No error

npx eslint test.ts
# No error

npx tsc
# `test.ts:3:3 - error TS18048: 'offset' is possibly 'undefined'.`

```
